# **Ruby Nouri Kermani - 30261323**

**Part 1: Data Cleaning and Preprocessing**

1.1 Load and Inspect the Dataset

In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


energy_data = pd.read_csv('/content/Building_Energy_Benchmarking.csv')
# loading the dataset

display(energy_data.shape)
print('\n')
display(energy_data.info())
print('\n')
display(energy_data.dtypes)
# diplaying dataset's shape, columns, and data types


(494, 31)



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 494 entries, 0 to 493
Data columns (total 31 columns):
 #   Column                                                                 Non-Null Count  Dtype  
---  ------                                                                 --------------  -----  
 0   Property Id                                                            494 non-null    int64  
 1   Property Name                                                          494 non-null    object 
 2   Address 1                                                              494 non-null    object 
 3   City                                                                   494 non-null    object 
 4   Postal Code                                                            494 non-null    object 
 5   Province                                                               494 non-null    object 
 6   Primary Property Type - Self Selected                                  494 non-null    o

None

Property Id                                                                int64
Property Name                                                             object
Address 1                                                                 object
City                                                                      object
Postal Code                                                               object
Province                                                                  object
Primary Property Type - Self Selected                                     object
Number of Buildings                                                        int64
Year Built                                                                 int64
Property GFA - Self-Reported (m²)                                         object
ENERGY STAR Score                                                        float64
Site Energy Use (GJ)                                                      object
Weather Normalized Site Energy Use (GJ)                                   object
Site EUI (GJ/m²)                                                         float64
Weather Normalized Site EUI (GJ/m²)                                      float64
Source Energy Use (GJ)                                                    object
Weather Normalized Source Energy Use (GJ)                                 object
Source EUI (GJ/m²)                                                       float64
Weather Normalized Source EUI (GJ/m²)                                    float64
Total GHG Emissions (Metric Tons CO2e)                                    object
Total GHG Emissions Intensity (kgCO2e/m²)                                float64
Direct GHG Emissions (Metric Tons CO2e)                                   object
Direct GHG Emissions Intensity (kgCO2e/m²)                               float64
Electricity Use - Grid Purchase (kWh)                                     object
Natural Gas Use (GJ)                                                      object
District Hot Water Use (GJ)                                               object
Electricity Use – Generated from Onsite Renewable Systems (kWh)          float64
Green Power - Onsite and Offsite (kWh)                                   float64
Avoided Emissions - Onsite and Offsite Green Power (Metric Tons CO2e)    float64
Year Ending                                                                int64
Unique ID                                                                 object
dtype: object

In [19]:
missing_values_engdata = energy_data.isnull().sum()
print(missing_values_engdata)
# finding the number of missing values in each column

Property Id                                                                0
Property Name                                                              0
Address 1                                                                  0
City                                                                       0
Postal Code                                                                0
Province                                                                   0
Primary Property Type - Self Selected                                      0
Number of Buildings                                                        0
Year Built                                                                 0
Property GFA - Self-Reported (m²)                                          0
ENERGY STAR Score                                                        329
Site Energy Use (GJ)                                                       0
Weather Normalized Site Energy Use (GJ)                                    0

1.2 Handling Missing Data

In [20]:
missing_values_percentage = (energy_data.isnull().sum() / len(energy_data)) * 100
print(missing_values_percentage)
print('\n')
# finding the percentage of columns

drop_columns = missing_values_percentage[missing_values_percentage > 40].index
energy_data_dropped = energy_data.drop(columns=drop_columns)
# dropping columns with more than 40% missing values

print(energy_data_dropped.isnull().mean()*100)
# just to make sure what columns with missing values are left


Property Id                                                               0.000000
Property Name                                                             0.000000
Address 1                                                                 0.000000
City                                                                      0.000000
Postal Code                                                               0.000000
Province                                                                  0.000000
Primary Property Type - Self Selected                                     0.000000
Number of Buildings                                                       0.000000
Year Built                                                                0.000000
Property GFA - Self-Reported (m²)                                         0.000000
ENERGY STAR Score                                                        66.599190
Site Energy Use (GJ)                                                      0.000000
Weat

In [26]:
# column = Weather Normalized Source Energy Use (GJ)
# type = object - categorical column

# column = Natural Gas Use (GJ)
# type = object - categorical column

# using mode / most frequece
energy_data_filled = energy_data_dropped.fillna({
    'Weather Normalized Source Energy Use (GJ)': energy_data_dropped['Weather Normalized Source Energy Use (GJ)'].mode(),
    'Natural Gas Use (GJ)': energy_data_dropped['Natural Gas Use (GJ)'].mode()
})
# filling the missing values of categorical columns with mode

from sklearn.impute import SimpleImputer

num_columns = energy_data_filled.select_dtypes(include=['float64', 'int64']).columns
imputer = SimpleImputer(strategy = 'median')
energy_data_filled[num_columns] = imputer.fit_transform(energy_data_filled[num_columns])
# filling the missing values of numerical columns with median

display(energy_data_filled)


Property Id                    Property Name                  Address 1  \
0     10176804.0  Acadia Aquatic & Fitness Centre       9009 Fairmount Dr SE   
1      6169481.0                       Ad Valorem              2924 11 ST NE   
2      6305956.0             Alberta Trade Centre               315 10 AV SE   
3      8854296.0                   Andrew Davison                133 6 AV SE   
4     21988603.0           Animal Services Centre        2201 PORTLAND ST SE   
..           ...                              ...                        ...   
489   21988624.0                  Senior Services               311 34 AV SE   
490   21988625.0                 Fire Station #25              4705 76 AV SE   
491   21988627.0                           EMS #6       1940 WESTMOUNT BV NW   
492   21988628.0                 Fire Station #24             2607 106 AV SW   
493   21988629.0                 Fire Station #36  10071 HIDDEN VALLEY DR NW   

        City Postal Code Province Primary Property Type - Self Selected  \
0    Calgary     T2H 0Z4  Alberta                  Heated Swimming Pool   
1    Calgary      t2e7l7  Alberta                                Office   
2    Calgary     T2G 0W2  Alberta                                Office   
3    Calgary      T2G0G2  Alberta                                Office   
4    Calgary     T2G 4M7  Alberta                                Office   
..       ...         ...      ...                                   ...   
489  Calgary     T2G 1T7  Alberta                                Office   
490  Calgary     T2C 2X1  Alberta                          Fire Station   
491  Calgary     T2N 3G8  Alberta               Other - Public Services   
492  Calgary     T2W 4H7  Alberta                          Fire Station   
493  Calgary     T3A 5G1  Alberta                          Fire Station   

     Number of Buildings  Year Built Property GFA - Self-Reported (m²)  ...  \
0                    1.0      2010.0                              1838  ...   
1                    1.0      1981.0                              7770  ...   
2                    1.0      1974.0                              6681  ...   
3                    1.0      1979.0                             17468  ...   
4                    1.0      2000.0                              2143  ...   
..                   ...         ...                               ...  ...   
489                  1.0      1960.0                               768  ...   
490                  1.0      1982.0                              1108  ...   
491                  1.0      1964.0                               425  ...   
492                  1.0      1978.0                              1108  ...   
493                  1.0      1988.0                               204  ...   

    Source EUI (GJ/m²) Weather Normalized Source EUI (GJ/m²)  \
0                 4.47                                  4.91   
1                 1.57                                  1.64   
2                 0.94                                  1.00   
3                 0.97                                  1.04   
4                 4.00                                  4.27   
..                 ...                                   ...   
489               1.57                                  1.64   
490               1.49                                  1.49   
491               2.27                                  2.37   
492               1.37                                  1.35   
493               2.05                                  2.02   

     Total GHG Emissions (Metric Tons CO2e)  \
0                                       509   
1                                       920   
2                                       499   
3                                      1293   
4                                       568   
..                                      ...   
489                                      66   
490                                     105   
49

1.3 Extracting and Cleaning Data Using Regex

**Part 2: Exploratory Data Analysis (EDA) and Aggregations**

2.1 Statistical Summary

2.2 Aggregations

2.3 Detecting Outliers Using Regex and IQR

**Part 3: Data Visualization**

3.1 Time-Series Visualization

3.2 Comparative Bar Charts

3.3 Heatmap Visualization

**Part 4: Further Analysis**

4.1 Correlation Analysis

4.2 Hypothesis Testing